pennylane/transforms/decorators.py

```diff
@@ -0,0 +1,388 @@
+
+    - The first argument must be an input tape
+
+    - Depending on the structure of this input tape, various quantum operations, functions,
+      and templates may be called.
+
+    - Any internal classical processing should use the ``qml.math`` module to ensure
+      the transform is differentiable.
+
+    - There is no return statement.
+
+    For example:
+
+    .. code-block:: python
+
+        @qml.single_tape_transform
+        def my_transform(tape, x, y):
+            # loop through all operations on the input tape
+            for op in tape.operations + tape.measurements:
+                if op.name == "CRX":
+                    wires = op.wires
+                    param = op.parameters[0]
+
+                    qml.RX(x * qml.math.abs(param), wires=wires[1])
+                    qml.RY(y * qml.math.abs(param), wires=wires[1])
+                    qml.CZ(wires=[wires[1], wires[0]])
+                else:
+                    op.queue()
+
+    This transform iterates through the input tape, and replaces any :class:`~.CRX` operation with
+    two single qubit rotations and a :class:`~.CZ` operation. These newly queued operations will
+    form the output transformed tape.
+
+    We can apply this transform to a quantum tape:
+
+    >>> with qml.tape.JacobianTape() as tape:
+    ...     qml.Hadamard(wires=0)
+    ...     qml.CRX(-0.5, wires=[0, 1])
+    >>> new_tape = my_transform(tape, 1., 2.)
+    >>> print(new_tape.draw())
+     0: ──H───────────────╭Z──┤
+     1: ──RX(0.5)──RY(1)──╰C──┤
+    """
+
+    def __init__(self, transform_fn):
+        self.transform_fn = transform_fn
+        functools.update_wrapper(self, transform_fn)
+
+    def __call__(self, tape, *args, **kwargs):
+        tape_class = type(tape.__class__.__name__, (NonQueuingTape, tape.__class__), {})
+
+        with tape_class() as new_tape:
+            self.transform_fn(tape, *args, **kwargs)
+
+        new_tape.__class__ = tape.__class__
+        return new_tape
+
+
+def qfunc_transform(tape_transform):
+    """Converts a single tape transform to a quantum function (qfunc) transform.
+
+    Args:
+        tape_transform (single_tape_transform): the single tape transform
+            to convert into the qfunc transform.
+
+    Returns:
+        function: A qfunc transform, that acts on any qfunc returning a *new*
+        qfunc as per the tape transform.
+
+    **Example**
+
+    Given a single tape transform ``my_transform(tape, x, y)``, you can use
+    this function to convert it into a qfunc transform:
+
+    >>> my_qfunc_transform = qfunc_transform(my_transform)
+
+    It can then be used to transform an existing qfunc:
+
+    >>> new_qfunc = my_qfunc_transform(old_qfunc, 0.6, 0.7)
+
+    It can also be used as a decorator:
+
+    .. code-block:: python
+
+        @qml.qfunc_transform
+        @qml.single_tape_transform
+        def my_transform(tape, x, y):
+            for op in tape.operations + tape.measurements:

```

Comment:
Why don't we need to construct a new tape here? (Similarly in the above example for tape transforms)

Reply:
It's automated! I realized, with both a single-tape transform and a qfunc transform, you are always transforming a single tape to another single tape. So it is boilerplate that will always be identical (I think-let me know if I am wrong!), and we can subsume it into the decorator to make it easier for developers.